WebHostLib/stats.py

```diff
@@ -0,0 +1,70 @@
+from collections import Counter, defaultdict
+from itertools import cycle
+from datetime import datetime, timedelta, date
+from math import pi
+
+from bokeh.embed import components
+from bokeh.palettes import Dark2_8 as palette
+from bokeh.plotting import figure, ColumnDataSource
+from bokeh.resources import INLINE
+from flask import render_template
+from pony.orm import select
+
+from . import app, cache
+from .models import Room
+
+pi2 = 2 * pi
+
+
+@app.route('/stats')
+@cache.memoize(timeout=60*60)  # regen once per hour should be plenty
+def stats():
```

Comment:
Is it possible to split this up so it's easier to add more charts in the future? Perhaps some list of classes which are responsible for building a figure based on a particular abstract base class? No reason this can't be more robust, imo.

Reply:
quite likely, though it's advisable to only go through and process the data once, then build figures from it, instead of accessing the DB for each figure's requirements. Does make the matter a bit more complex. Could certainly split the function into sections with some comment headers.

Reply:
You can preload the data in stats() then process through the registered FigureBuilders (example name).